src/archetype.rs

```diff
@@ -26,6 +26,7 @@ use crate::{Access, Component, Query};
 /// [`World`](crate::World).
 pub struct Archetype {
     types: Vec<TypeInfo>,
+    type_ids: Box<[TypeId]>,
```

Comment:
This seems a bit of a price to pay to produce a matching slice of `TypeId`. Maybe the bundle interface can actually make due with `(ExactSize)Iterator` so that `Archetype::types` could also fulfil the requirement?

Reply:
My feeling is that incremental fixed costs on archetype construction are not a significant overhead, since archetype construction is already relatively expensive.

`with_ids` is only called in connection with `ArchetypeSet::get`, which passes it to `HashMap::<Box<[TypeId]>, _>::get`. Substituting an iterator there is awkward, likely requiring replacing the `HashMap` with a [`RawTable`](https://docs.rs/hashbrown/latest/hashbrown/raw/struct.RawTable.html).

Given the uncertain cost and complexity of mitigating it, I am inclined to deem this a premature opt.

Reply:
> Given the uncertain cost and complexity of mitigating it, I am inclined to deem this a premature opt.

Agreed. I was not thinking far enough and using `Iterator` would just mess things up elsewhere.

One idea related to bundles even if unrelated to this PR would be to modify the `DynamicBundle` trait to be more operation oriented, e.g. instead of providing slices of `TypeId` or `TypeInfo`, it could get methods for performing more specific operations like "check whether this bundle contains `T`?" or "here is a set of type ID, insert all type ID which are part of you but not yet part of the set". This would probably make implementing the trait more involved, but also more flexible as to the shape of the data backing those operations.

Reply:
That's an intriguing prospect, though indeed best pursued in future work. It might help rein in the complexity/ugliness of the traits, perhaps even to the point that we could make them public. Some additional `World` and/or `Archetype` APIs would likely need to be defined for implementations to call into.

Reply:
Opened https://github.com/Ralith/hecs/issues/249 to track that.